### PR TITLE
Added new fields to context output and UI

### PR DESCRIPTION
**Pull request by jlevypaloalto**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: [PAN-OS command pan-os-list-rules doesn't return all data](https://jira-hq.paloaltonetworks.local/browse/CIAC-6143)

## Description
Addresses the issue in the PAN-OS integration where the `pan-os-list-rules` command doesn't provide all the required data to the context. This pull request adds missing data like "Security Profile" and "Options" to the command output. Additionally, the response now indicates whether a profile is a Security Profile or a Security Profile Group.

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (JudahSchwartz) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:3646]: why take out location if not panorama? If its a matter of not having null columns removeNull should do this job

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (JudahSchwartz) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:3652]: same comment here for target

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (JudahSchwartz) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:3449]: we have dict_safe_get in CSP, lets use instead

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (JudahSchwartz) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:3527]: we dont need to maintain 2 dicts here, as the pretty rules are a subset of the full rules. We should be limiting the human readable by using the chosen headers, not giving a seperate dict

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (jlevypaloalto) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:3646]: I wanted the readable outputs to mirror their respective UIs
should i just have one for both?

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (jlevypaloalto) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:3527]: Thats what i wanted to do originally but pretty rules is not an exact subgroup
pretty rules mimics the UI and context rules cant break backwards compatibility so their key are different

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (jlevypaloalto) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:3449]: although i found some other places where i can use that

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:3478]: why do you need all of this?

you can use the `parse_pan_os_un_committed_data` which does all of this for you.

what's the purpose of this function? if i understand its to get data from un-committed data?

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:11581]: why did you delete this?

**Review by GuyAfik — COMMENTED** [review]
I think using the `parse_pan_os_un_committed_data` can save you some of the code you have wrote.

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (jlevypaloalto) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:11581]: its a duplicate

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (JudahSchwartz) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py]: which key are you talking about? Human readable shouldnt be different than the context for the most part

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (JudahSchwartz) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py]: it does work recursively, 
for example

    `alert_csv_id = dict_safe_get(alert, ['alert_data', 'csv', 'id'])`
check in the code how it is used

Comment (jlevypaloalto) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py]: I should have been more clear. heres an example:

in one rule you have:
```
"destination-hip": {
                        "@admin": "demistoadmin",
                        "@dirtyId": "12",
                        "@time": "2023/05/30 03:17:35",
                        "member": {
                            "@admin": "demistoadmin",
                            "@dirtyId": "12",
                            "@time": "2023/05/30 03:17:35",
                            "#text": "any",
                        },
                    }
```

and in another:

`"destination-hip": {"member": "any"}`

i need to get the value "any" regardless of how many keys deep it is
this function will dig until it gets a value and not a dict

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (jlevypaloalto) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py]: for example:
` context_rule['To'] = pretty_rule['Destination Zone'] = rule_get('to')`
there are better examples at the beginning of the function.
Ido stressed in the demo that he wanted the HR to look like the UI so thats why i did it, it also makes more sense because with all the new fields we're adding the old version's fields would be to vague

Comment (JudahSchwartz) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py]: lets use a headerTransform function for this then instead. You can have a dict of from_header_name to to_header_name.

Comment (jlevypaloalto) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py]: the good news is that the keys are the same object in memory so we're not waisting to much space

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (JudahSchwartz) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:3486]: @GuyAfik does this seem like a good function to use in panos? Is the complexity warranted?
If yes lets write very clear docstrings and tests

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (JudahSchwartz) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py]: the concern here isnt space, its readability and needless complexity

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:3570]: pretty hard to read it like this, can you align all the lines to avoid it?

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:3520]:
```suggestion
    if isinstance(value, dict):
```

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:3515]:
```suggestion
    if not isinstance(xml_dict, dict):
```

please use `isinstance` instead of `is` in all places

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:3549]: what is this?

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:3565]:
```suggestion
    context_rule['DeviceGroup'] = DEVICE_GROUP
```

`None` not needed, its empty string by default

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:3569]: very hard to read, please align this to be more clear into a single line

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:3580]: use isinstance

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:3585]: align the line

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:3591]: please explain

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:3595]: use `get` of dict instead of this, can fail on `KeyError` if it doesn't exist, lets prevent that

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:3623]: use f-strings, isinstance function instead of `type` and align the lines as much as possible to be more readable

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:11592]: nice catch

File: Packs/PAN-OS/Integrations/Panorama/Panorama.yml
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.yml:5798]: why?
you realize validation will fail on it? its backward compatibility break

File: Packs/PAN-OS/ReleaseNotes/1_17_4.md
Comment (GuyAfik) [inline Packs/PAN-OS/ReleaseNotes/1_17_4.md:8]:
```suggestion
```

u didn't update the dockerimage at all

File: Packs/PAN-OS/ReleaseNotes/1_17_4.md
Comment (GuyAfik) [inline Packs/PAN-OS/ReleaseNotes/1_17_4.md:7]:
```suggestion
- Added additional context-output to the ***pan-os-list-rules*** command. 
```

**Review by GuyAfik — CHANGES_REQUESTED** [review]
@jlevypaloalto nice work! please see my comments :)

**Comment by GuyAfik** [comment]
@jlevypaloalto please update the PR description to something more meaningful

File: Packs/PAN-OS/Integrations/Panorama/Panorama.yml
Comment (GuyAfik) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.yml:5798]: then you will need a force merge......

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:3610]: use f string

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:3613]: pretty hard to read IMO, can we make it a bit simpler?

File: Packs/PAN-OS/ReleaseNotes/1_17_4.md
Comment (GuyAfik) [reply Packs/PAN-OS/ReleaseNotes/1_17_4.md:7]: ?

**Review by GuyAfik — COMMENTED** [review]
@jlevypaloalto looks good, @JudahSchwartz please review it yourself too

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (JudahSchwartz) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:3549]: if we arent using these kinds of checkmarks in the rest of the integration, lets not do it here

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py]: from what i understand, the purpose here is to extract nested fields from the response, there are advantages which makes it easier to handle nested values recursively, however IMO, it makes the readability a bit harder and we want it to be as easy as possible to understand in case we would have to deal with bugs in the future.

from my point of view i don't see any harm as long as we can generalize it to other cases / other XML api's which might use it. 

We should definitely add very good unit-tests here to cover as many cases as possible.

Comment (JudahSchwartz) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py]: lets try to write this function clearer as well. it shouldnt be recursive

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (jlevypaloalto) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:3549]: et tu Judah? then fall tick_box

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (jlevypaloalto) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:3610]: np

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (jlevypaloalto) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:3613]: on it

File: Packs/PAN-OS/Integrations/Panorama/Panorama.yml
Comment (jlevypaloalto) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.yml:5798]: i will anyway because of the "unsafe xml" thing

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (jlevypaloalto) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py]: ill work on making the function better

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (jlevypaloalto) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py]: Great tip, will use headerTransform and one dict

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (jlevypaloalto) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py]: sure np

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (jlevypaloalto) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py]: Changed my mind. Im not using f string 😉

File: Packs/PAN-OS/Integrations/Panorama/README.md
Comment (JudahSchwartz) [inline Packs/PAN-OS/Integrations/Panorama/README.md:3204]: lets not use emojis

Comment (jlevypaloalto) [reply Packs/PAN-OS/Integrations/Panorama/README.md:3204]: np

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (JudahSchwartz) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:3643]: you can make the call site  `lambda x: to_human_readable.get(x, x)` and then leave everything out of the dict that doesnt need to be changed

Comment (JudahSchwartz) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:3643]: example Packs.CrowdStrikeFalconSandbox.Integrations.CrowdStrikeFalconSandboxV2.CrowdStrikeFalconSandboxV2.crowdstrike_analysis_overview_summary_command

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:3473]:
```suggestion
    rule_devices = [
        entry.get('@name') for entry in entries if isinstance(entry, dict)
    ]
```

**Review by GuyAfik — COMMENTED** [review]
@jlevypaloalto very nice, I liked your idea, some minor comments and its good to go IMO.

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (JudahSchwartz) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:3475]: lets call this context_rule as it isnt just for hr

**Review by JudahSchwartz — APPROVED** [review]
looks good to me. Please wait for @GuyAfik's approval

**Review by GuyAfik — APPROVED** [review]
nice!

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (jlevypaloalto) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:3475]: that was the original name and its used everwere, let me know if its critical

**Pull request merged**
